receiver/receivertest/contract_checker.go

```diff
@@ -0,0 +1,405 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package receivertest // import "go.opentelemetry.io/collector/receiver/receivertest"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.opentelemetry.io/collector/receiver"
+)
+
+// UniqueIDAttrName is the attribute name that is used in log records/spans/datapoints as the unique identifier.
+const UniqueIDAttrName = "test_id"
+
+// UniqueIDAttrDataType is the value type of the UniqueIDAttrName.
+type UniqueIDAttrDataType int64
+
+type Generator interface {
+	// Generate must generate and send at least one data element (span, log record or metric data point)
+	// to the receiver and return a copy of generated element ids.
+	// The generated data must contain uniquely identifiable elements, each with a
+	// different value of attribute named UniqueIDAttrName.
+	// CreateOneLogWithID() can be used a helper to create such logs.
+	Generate() IDSet
+}
+
+type CheckConsumeContractParams struct {
+	T *testing.T
+	// Factory that allows to create a receiver.
+	Factory receiver.Factory
+	// DataType to test for.
+	DataType component.DataType
+	// Config of the receiver to use.
+	Config component.Config
+	// Generator that can send data to the receiver.
+	Generator Generator
+	// GenerateCount specifies the number of times to call the generator.Generate().
+	GenerateCount int
+}
+
+// CheckConsumeContract checks the contract between the receiver and its next consumer. For the contract
+// description see ../doc.go. The checker will detect violations of contract on different scenarios: on success,
+// on permanent and non-permanent errors and mix of error types.
+func CheckConsumeContract(params CheckConsumeContractParams) {
```

Comment:
I tried this on filelog receiver in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/22833 and run into problems with reusing the same generator and config across multiple tests.
I fixed it by changing this signature to access a function returning `CheckConsumeContractParams`, but I don't think it's the best idea. Another, maybe a better, option is to accept a generator factory instead of the generator itself. Then, I think we need an additional `Stop()` method on it.

Otherwise, it worked pretty well. Thank you for adding these tests!

Reply:
Can you clarify what the problem is with reusing the config? Do you need to have different configs for different subtests? If you don't need different configs we can add Start() method to Generator, where you can re-create the test input file.

Reply:
I added Start() to Generator. See if it helps the filelog use case.

Reply:
The problem is we create a new receiver for every subtest, in which I need to set up a file to send the logs through. I have two options:
1. The first option is to use a single file, pass it to the filelog configuration, and append data to the file for each subtest on top of the existing data. This approach would be feasible if the filelog receiver could be configured to read from the end of the same file in every test case. However, the filelog receiver begins polling files only after [a certain interval](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3fb4d9963cc5e8f01c0f7c6894b1248f48162396/pkg/stanza/fileconsumer/file.go#L87-L95), by which time all the logs have already been written by the generator.
2. The second option is to generate a new file for each subtest in the `Generator.Start()` method and set up the filelog receiver to read from the beginning of all files that match a specific file pattern. This option would work, but I need to be able to remove the file after each subtest is finished. Otherwise, subsequent subtests would end up reading data from files created for previous subtests.

I would go with the (2), but I need `Generator.Stop()` or `Shutdown()` method additionally to `Start()`.

Reply:
Sounds good, I think it is reasonable for Generator to have a Start/Stop. I will add.

Reply:
Added Stop().